draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+As part of a [presentation](https://github.com/troessner/talks/blob/master/exploiting_rubys_ast_a_love_story_in_three_chapters/presentation.html) I'm hoping to give at the end of this year about abstract syntax trees and how you can leverage them I started to look into how the [Mutant](https://github.com/mbj/mutant) gem works. `Mutant` is the by far most advanced gem for doing mutation testing in Ruby. I was (and still am) amazed how incredibly well built it is on all levels, ranging from the "big picture" architecture down to the "nuts and bolts" source code.
+
+When it comes to mutation testing I learned one thing. It's complex. Really complex. And as a consequence, `Mutant` is complex. 
```

Comment:
I actually do not agree to the fact that mutation testing is complex in itself. Its a pretty easy to describe technique. But the implementation mirrors the complexity of the environment its applied to.

Lots of the complexity mutant exposes, is just mirroring the inner complexity of ruby as a language and its ecosystem.

Other languages with better inner workings, better tooling, and as effect less hackish tooling (tooling on ruby is more hackish as the lanugage makes it easy to cut corners) do not surface that much complexity.